File: test/EulerStepCounter_test.cpp
#include "EulerStepCounter.h"
#include "gtest/gtest.h"

TEST(EulerStepCounterTest, clockwiseRotation) {
  EulerStepCounter instance;

  instance.update(0.2);
  ASSERT_FLOAT_EQ(0.2, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(0.4, instance.linear_position());

  instance.update(0.6);
  ASSERT_FLOAT_EQ(0.6, instance.linear_position());

  instance.update(0.89);
  ASSERT_FLOAT_EQ(0.89, instance.linear_position());

}

TEST(EulerStepCounterTest, clockwiseRotationWithFullRotation) {
  EulerStepCounter instance;

  instance.update(0.2);
  ASSERT_FLOAT_EQ(0.2, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(0.4, instance.linear_position());

  instance.update(0.6);
  ASSERT_FLOAT_EQ(0.6, instance.linear_position());

  instance.update(0.89);
  ASSERT_FLOAT_EQ(0.89, instance.linear_position());

  instance.update(0.2);
  ASSERT_FLOAT_EQ(1.2, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(1.4, instance.linear_position());
}


TEST(EulerStepCounterTest, clockwiseRotationWithDirectionChange) {
  EulerStepCounter instance;

  instance.update(0.1);
  ASSERT_FLOAT_EQ(0.1, instance.linear_position());

  instance.update(0.3);
  ASSERT_FLOAT_EQ(0.3, instance.linear_position());

  instance.update(0.6);
  ASSERT_FLOAT_EQ(0.6, instance.linear_position());

  instance.update(0.8);
  ASSERT_FLOAT_EQ(0.8, instance.linear_position());

  instance.update(0.2);
  ASSERT_FLOAT_EQ(1.2, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(1.4, instance.linear_position());

  instance.update(0.7);
  ASSERT_FLOAT_EQ(1.7, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(1.4, instance.linear_position());

  instance.update(0.3);
  ASSERT_FLOAT_EQ(1.3, instance.linear_position());

  instance.update(0.1);
  ASSERT_FLOAT_EQ(1.1, instance.linear_position());
}

TEST(EulerStepCounterTest, handlesReverseDirection) {
  EulerStepCounter instance;

  instance.update(0.1);
  ASSERT_FLOAT_EQ(0.1, instance.linear_position());

  instance.update(0.3);
  ASSERT_FLOAT_EQ(0.3, instance.linear_position());

  instance.update(0.6);
  ASSERT_FLOAT_EQ(0.6, instance.linear_position());

  instance.update(0.8);
  ASSERT_FLOAT_EQ(0.8, instance.linear_position());

  instance.update(0.2);
  ASSERT_FLOAT_EQ(1.2, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(1.4, instance.linear_position());

  instance.update(0.7);
  ASSERT_FLOAT_EQ(1.7, instance.linear_position());

  instance.update(0.4);
  ASSERT_FLOAT_EQ(1.4, instance.linear_position());

  instance.update(0.3);
  ASSERT_FLOAT_EQ(1.3, instance.linear_position());

  instance.update(0.1);
  ASSERT_FLOAT_EQ(1.1, instance.linear_position());

  instance.update(0.3);
  ASSERT_FLOAT_EQ(1.3, instance.linear_position());

  instance.update(0.7);
  ASSERT_FLOAT_EQ(1.7, instance.linear_position());

  instance.update(0.9);
  ASSERT_FLOAT_EQ(1.9, instance.linear_position());

  instance.update(0.1);
  ASSERT_FLOAT_EQ(2.1, instance.linear_position());

  instance.update(0.8);
  ASSERT_FLOAT_EQ(1.8, instance.linear_position());
}

File: euler_step_counter_test.rb
require 'test/unit'
require_relative 'euler_step_counter'

class EulerStepCounterTest < Test::Unit::TestCase
  def setup
    @counter = EulerStepCounter.new
  end

  def test_initializes_attributes
    assert_equal 0.0, @counter.current_angle
    assert_equal 0.0, @counter.previous_angle
    assert_equal 0.0, @counter.diff
    assert_equal 0.0, @counter.rotations
  end

  def test_linear_position_on_clockwise_rotation
    @counter.update(0.2)
    assert_equal 0.2, @counter.linear_position

    @counter.update(0.4)
    assert_equal 0.4, @counter.linear_position

    @counter.update(0.6)
    assert_equal 0.6, @counter.linear_position

    @counter.update(0.89)
    assert_equal 0.89, @counter.linear_position
  end

  def test_linear_position_on_clockwise_transition_is_incremental
    @counter.update(0.2)
    assert_equal 0.2, @counter.linear_position

    @counter.update(0.4)
    assert_equal 0.4, @counter.linear_position

    @counter.update(0.6)
    assert_equal 0.6, @counter.linear_position

    @counter.update(0.89)
    assert_equal 0.89, @counter.linear_position

    @counter.update(0.1)
    assert_equal 1.1, @counter.linear_position

    @counter.update(0.3)
    assert_equal 1.3, @counter.linear_position
  end

  def test_linear_position_with_change_of_direction
    @counter.update(0.1)
    assert_equal 0.1, @counter.linear_position

    @counter.update(0.3)
    assert_equal 0.3, @counter.linear_position

    @counter.update(0.6)
    assert_equal 0.6, @counter.linear_position

    @counter.update(0.8)
    assert_equal 0.8, @counter.linear_position

    @counter.update(0.2)
    assert_equal 1.2, @counter.linear_position

    @counter.update(0.4)
    assert_equal 1.4, @counter.linear_position

    @counter.update(0.7)
    assert_equal 1.7, @counter.linear_position

    @counter.update(0.4)
    assert_equal 1.4, @counter.linear_position

    @counter.update(0.3)
    assert_equal 1.3, @counter.linear_position

    @counter.update(0.1)
    assert_equal 1.1, @counter.linear_position

    @counter.update(0.8)
    assert_equal 2.8, @counter.linear_position
  end

  def test_handles_anti_clockwise_start
    @counter.update(0.8)
    assert_equal 0.8, @counter.linear_position
  end
end
